src/redirect.rs

```diff
@@ -0,0 +1,125 @@
+use anyhow::{anyhow, Result};
+use reqwest::blocking::Client;
+use reqwest::blocking::{Request, Response};
+use reqwest::header::{
+    HeaderMap, AUTHORIZATION, CONTENT_ENCODING, CONTENT_LENGTH, CONTENT_TYPE, COOKIE, LOCATION,
+    PROXY_AUTHORIZATION, TRANSFER_ENCODING, WWW_AUTHENTICATE,
+};
+use reqwest::{Method, StatusCode, Url};
+
+pub struct RedirectFollower<'a, T>
+where
+    T: FnMut(Response, &mut Request) -> Result<()>,
+{
+    client: &'a Client,
+    max_redirects: usize,
+    callback: Option<T>,
+}
+
+impl<'a, T> RedirectFollower<'a, T>
+where
+    T: FnMut(Response, &mut Request) -> Result<()>,
+{
+    pub fn new(client: &'a Client, max_redirects: usize) -> Self {
+        RedirectFollower {
+            client,
+            max_redirects,
+            callback: None,
+        }
+    }
+
+    pub fn on_redirect(&mut self, callback: T) {
+        self.callback = Some(callback);
+    }
+
+    pub fn execute(&mut self, mut first_request: Request) -> Result<Response> {
+        let mut request = clone_request(&mut first_request);
+        let mut response = self.client.execute(first_request)?;
```

Comment:
I think the whole buffering conundrum is worth a comment:
```suggestion
        // This buffers the body in case we need it again later
        // reqwest does *not* do this, it ignores 307/308 with a streaming body
        let mut request = clone_request(&mut first_request);
        let mut response = self.client.execute(first_request)?;
```